### PR TITLE
feat(app): better upgrade management

### PR DESCRIPTION
**Pull request by technicallyty**
## Description

implements better upgrade management like [Osmosis](https://github.com/osmosis-labs/osmosis/tree/main/app/upgrades). 

Benefits:
- easier to remember to add store upgrades
- can directly test an upgrade without manually setting up an app
- nice way to organize all upgrades without clogging up one giant file

Cons:
- uses the binary the app will upgrade *to*, which can create odd situations when we do things like force a module's from version to be its consensus version (skipping its InitGenesis). 
- the `CreateUpgradeHandler` isn't as extensible as one might like, for example if they require a reference to the app. For most cases, the current function should suffice.

Closes: #1440

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: app/appconfig.go [hunk 1]
```diff
@@ -1,84 +1,12 @@
 package app
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply app/appconfig.go:1]: On second thought, can we just create the new ante handler within `app.go` ? I don't think we need this to be separate and this private function is only used once in `app.go`.

File: app/apptesting/test_suite.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+type KeeperTestHelper struct {
+	suite.Suite
+
+	App         *app.RegenApp
+	Ctx         sdk.Context
+	QueryHelper *baseapp.QueryServiceTestHelper
+	TestAccs    []sdk.AccAddress
+}
```
Comment (ryanchristo) [inline app/apptesting/test_suite.go:23]: I realize this is what osmosis has but this seems more like an `AppTestSuite` than a `KeeperTestHelper`. I don't think we should copy over the naming for this.

File: app/apptesting/test_suite.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// CreateRandomAccounts is a function return a list of randomly generated AccAddresses
+func CreateRandomAccounts(numAccts int) []sdk.AccAddress {
+	testAddrs := make([]sdk.AccAddress, numAccts)
+	for i := 0; i < numAccts; i++ {
+		pk := ed25519.GenPrivKey().PubKey()
+		testAddrs[i] = sdk.AccAddress(pk.Address())
+	}
+
+	return testAddrs
+}
```
Comment (ryanchristo) [inline app/apptesting/test_suite.go:46]: This should probably go in a separate file with the other "test helpers". I think we should migrate `test_helpers.go` to a `testutil` package and include this with the other methods that are not directly related to the app test suite.

File: app/test_helpers.go [hunk 4]
```diff
@@ -54,6 +54,18 @@ type SetupOptions struct {
...
+	return SetupOptions{
+		Logger:             log.NewNopLogger(),
+		DB:                 dbm.NewMemDB(),
+		InvCheckPeriod:     5,
+		HomePath:           DefaultNodeHome,
+		SkipUpgradeHeights: map[int64]bool{},
+		EncConfig:          MakeEncodingConfig(),
+		AppOpts:            EmptyAppOptions{},
+	}
+}
```
Comment (ryanchristo) [inline app/test_helpers.go:67]: As mentioned above, I think these are all app test utils and should live together in a single package.

**Review by ryanchristo — COMMENTED** [review]
Nice! While we are reorganizing a bit, I have a few suggestions. I don't think we should copy the structure of osmosis in some cases and we can consolidate test utils into a single package.

File: app/appconfig.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryanchristo) [reply app/appconfig.go:1]: We can also handle in a separate pull request but renaming `setCustomAnteHandler` to `anteHandler` got me thinking.

File: app/test_helpers.go (hunk 4 above)
Comment (ryanchristo) [reply app/test_helpers.go:67]: We can also handle migrating these helpers in a separate pull request but we should prepare the `testutil` package if you think this would be a good idea.

File: app/apptesting/test_suite.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package apptesting
```
[the remark this answers is not in the document]
Comment (technicallyty) [reply app/apptesting/test_suite.go]: [chore: rename apptesting -> testutil](https://github.com/regen-network/regen-ledger/pull/1488/commits/466889ff20aad2cdd513426ca3dee58aae3d47ce)

File: app/appconfig.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (technicallyty) [reply app/appconfig.go]: [chore: move ante handler in to app.go](https://github.com/regen-network/regen-ledger/pull/1488/commits/dd1560f1f68bc7b76ffec8c4d188341c0d403a31) just put in in app.go, can always break out later if we have more ante handlers stuff to add

File: app/apptesting/test_suite.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (technicallyty) [reply app/apptesting/test_suite.go]: this caused some import cycle errors, will fix in morning

[the remark this answers is not in the document]
Comment (technicallyty) [reply app/apptesting/test_suite.go]: ok should be fixed, move app_test to the testutil package

**Review by ryanchristo — APPROVED** [review]
[tACK](https://hackmd.io/16fHDJw8Q3-872ivCLMnZQ).

It's not entirely clear to me why osmosis keeps the upgrade handlers for each upgrade. As far as I can tell, this is the only cosmos project that does this. Adding an upgrade test and separating out the upgrade logic into a separate package is a nice addition but subpackages for each upgrade seems unnecessary to me and will lead to the accumulation of dead code.

For example, a `v5` binary does not need and will never make use of a `v4` upgrade handler (or any previous upgrade handlers). The `v4` upgrade handler only needs to be included in the `v4` binary. If a chain were upgrading from `v3` to `v5`, node operators would still need to switch to the `v4` binary that includes the `v4` upgrade handler and then they would need to switch to the `v5` binary that includes the `v5` upgrade handler.

The sdk will be switching to proper semantic versioning for breaking changes soon but until that happens we may also run into a situation where we would need to make changes to a previous upgrade handler and then it seems we would be maintaining dead code. Given we are only moving the `v5` upgrade handler and not adding previous upgrade handlers, I think this is ok for now but maybe worth discussing.

**Comment by aleem1314** [comment]
> If a chain were upgrading from `v3` to `v5`, node operators would still need to switch to the `v4` binary that includes the `v4` upgrade handler and then they would need to switch to the `v5` binary that includes the `v5` upgrade handler.

A chain can upgrade from v3 -> v5 without v4 binary. Module manager first runs `v3` -> `v4` migration and then `v4` -> `v5` migration.

**Review by aleem1314 — APPROVED** [review]
Lgtm
